### PR TITLE
fix: Clean up VAT displaying condition in document templates

### LINKED ISSUE #8571: Clean up VAT displaying condition in document templates
### Technical TODO

Clean up and simplify VAT displaying condition in the document templates [letter_header.html.twig](https://github.com/shopware/shopware/blob/4698d5970b0bdb36a11f18b147b42d32f141bda5/src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig#L36-L38) (line 36-38) and [position_header.html.twig](https://github.com/shopware/shopware/blob/39f86f4b064814f30b103484c447ce78157a68a6/src/Core/Framework/Resources/views/documents/includes/position_header.html.twig#L36-L42) (line 36-42)

Clarify what is the correct condition and update the template accordingly. Currently billing address is checked but maybe that is wrong and it should be shipping. Preferably add a new template variable separately for it.

### PR DESCRIPTION
**Pull request by DennisGarding**
Closes: #8571

### 1. Why is this change necessary?

VAT information used outdated conditions.

### 2. What does this change do, exactly?

Uses intraCommunityDelivery consistently and adds test coverage.

### 3. Describe each step to reproduce the issue or behavior.

Generate an intra-community invoice and check its VAT information.

### CODE REVIEW

File: src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig [hunk 1]
```diff
@@ -28,8 +29,11 @@ All blocks of this template are available in the template which renders this tem
                     address: billingAddress
                 } %}<br/>
 
-                {% if (customer.customer.vatIds|first and config.displayCustomerVatId) or (customer.customer.vatIds|first and billingAddress.country.companyTax.enabled and config.displayAdditionalNoteDelivery and billingAddress.country.id in config.deliveryCountries) %}
-                    {{ 'document.vatId'|trans({'%vatId%': customer.customer.vatIds|first})|sw_sanitize(null, true) }}<br>
+                {% set customerVatId = customer.customer.vatIds|first -%}
+                {% set displayCustomerVatId = customerVatId and (config.displayCustomerVatId or config.intraCommunityDelivery) -%}
```
Comment (CR0YD) [inline src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig:33]:
```suggestion
                {% set customerVatId = customer.customer.vatIds|first %}
                {% set displayCustomerVatId = customerVatId and (config.displayCustomerVatId or config.intraCommunityDelivery) %}
```
Or are `-` at the end necessary?

Comment (DennisGarding) [reply src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig:33]: This is Twig-Whitespace-Control.

This causes Twig to remove all immediately following spaces and line breaks up to the next piece of content.

See: http://twig.symfony.com/doc/3.x/templates.html#whitespace-control

Comment (CR0YD) [reply src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig:33]: Ah okay. Thanks for pointing that out!

**Review by CR0YD — DISMISSED** [review]

File: src/Core/Framework/Resources/views/documents/includes/position_header.html.twig [hunk 2]
```diff
@@ -39,7 +40,7 @@ All blocks of this template are available in the template which renders this tem
                         {% set taxStatusGross = order.price.taxStatus == constant('Shopware\\Core\\Checkout\\Cart\\Price\\Struct\\CartPrice::TAX_STATE_GROSS') %}
                         {% set taxStatusNet = order.price.taxStatus == constant('Shopware\\Core\\Checkout\\Cart\\Price\\Struct\\CartPrice::TAX_STATE_NET') %}
 
-                        {% set displayVAT = not companyTaxEnabled or not (displayAdditionalNoteDelivery and isDeliveryCountry) %}
+                        {% set displayVAT = not config.intraCommunityDelivery %}
```
Comment (MalteJanz) [inline src/Core/Framework/Resources/views/documents/includes/position_header.html.twig]: This new condition looks quite a bit different than the one before, looking at its new definition:
https://github.com/shopware/shopware/blob/5f8c15c58f82fead8308ab714a5e326cbd2e6d08/src/Core/Checkout/Document/Renderer/AbstractDocumentRenderer.php#L53-L84

Not sure if we should apply this as a bug fix or as a behavior break with the next major 🤔 . For example its now based on the shipping address instead of the billing address. ~~Additionally our upcoming new doc generation implementation goes even further and also validates the VAT IDs~~ (actually the old implementation does as well):
https://github.com/shopware/shopware/blob/427b55bb5f129071fe2da9be34e92dd4a3c6c712/src/Core/Checkout/DocumentV2/Provider/InvoiceDataProvider.php#L174-L223

I'm also unsure if we want to mix the two concepts of "intraCommunityDelivery" and "VAT" like this, aren't they two different things? Where one (intraCommunityDelivery) has impact on the other (VAT), e.g. prevents it from displaying, but couldn't there be other cases where we don't want to display VAT that aren't "intraCommunityDelivery" type orders?

My suggestion (feel free to suggest something else):
1. Make this a major change, behind our document generation refactor feature flag, so we don't suddenly break some behavior in a minor version and potentially cause issues for merchants and customers
2. We likely want to do some more research and add more test cases to ensure the new condition / logic is legally correct for all cases, maybe @timshopware can help with that? Related discussion in another issue: https://github.com/shopware/shopware/issues/13618#issuecomment-3943543375

**Review by MalteJanz — CHANGES_REQUESTED** [review]

File: src/Core/Framework/Resources/views/documents/includes/position_header.html.twig (hunk 2 above)
Comment (DennisGarding) [reply src/Core/Framework/Resources/views/documents/includes/position_header.html.twig]: ok, i have added the fix behind the feature flag and added more tests. 
To correct the intra-community-label behavior, we should wait for a decision and create a new ticket for this.

File: RELEASE_INFO-6.7.md [hunk 3]
```diff
@@ -6,6 +6,9 @@
 
 ## Core
 
+### Document templates use the DocumentV2 VAT display condition behind the 6.8 feature flag
+
+The document templates now use the shipping-based `intraCommunityDelivery` condition for displaying VAT information when the `v6.8.0.0` feature flag is active, matching the DocumentV2 document generation path. With the feature flag disabled, the existing billing-address and configured delivery-country behavior remains unchanged. Extensions and themes that render or assert the legacy document templates should test their output with the feature flag enabled before upgrading to Shopware 6.8.
```
Comment (larskemper) [inline RELEASE_INFO-6.7.md:59]:
```suggestion
The document templates now use the shipping-based `intraCommunityDelivery` condition for displaying VAT information when the `v6.8.0.0` feature flag is active, matching the DocumentV2 document generation path. With the feature flag disabled, the existing billing-address and configured delivery-country behavior remains unchanged. Extensions and themes that render or assert the legacy document templates should test their output with the feature flag enabled before upgrading to Shopware 6.8.

```

**Review by larskemper — DISMISSED** [review]

File: src/Core/Framework/Resources/views/documents/includes/position_header.html.twig (hunk 2 above)
Comment (MalteJanz) [reply src/Core/Framework/Resources/views/documents/includes/position_header.html.twig]: > ok, i have added the fix behind the feature flag

To be completely honest, I think that made the code even worse (adding the feature check to the existing conditions). The goal of the issue was to simplify this mess, not make it worse 😬.

Not sure if its possible, but you could try to wrap the `{% set ... %}` declarations in an `{% if feature('v6.8.0.0') %}`, then its at least more readable on how the condition looks like now and how it will look like in the major. According to the twig docs, the set instruction is only scoped to for loops, not sure if that applies to if statemetns as well:
https://twig.symfony.com/doc/3.x/tags/set.html

Comment (DennisGarding) [reply src/Core/Framework/Resources/views/documents/includes/position_header.html.twig]: I have simplified the condition

File: src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig [hunk 4]
```diff
@@ -28,8 +29,11 @@ All blocks of this template are available in the template which renders this tem
                     address: billingAddress
                 } %}<br/>
 
-                {% if (customer.customer.vatIds|first and config.displayCustomerVatId) or (customer.customer.vatIds|first and billingAddress.country.companyTax.enabled and config.displayAdditionalNoteDelivery and billingAddress.country.id in config.deliveryCountries) %}
-                    {{ 'document.vatId'|trans({'%vatId%': customer.customer.vatIds|first})|sw_sanitize(null, true) }}<br>
+                {% set customerVatId = customer.customer.vatIds|first -%}
+                {% set displayCustomerVatId = customerVatId and (config.displayCustomerVatId or (feature('v6.8.0.0') and config.intraCommunityDelivery)) -%}
+
+                {% if displayCustomerVatId or (not feature('v6.8.0.0') and customerVatId and billingAddress.country.companyTax.enabled and config.displayAdditionalNoteDelivery and billingAddress.country.id in config.deliveryCountries) %}
```
Comment (MalteJanz) [inline src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig]: nit: is it also possible to simplify these ones? The other file looks much better now 👍

**Review by MalteJanz — DISMISSED** [review]

File: src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig (hunk 4 above)
Comment (DennisGarding) [reply src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig]: It is now also more easy

File: tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php [hunk 5]
```diff
@@ -105,6 +105,8 @@ protected function tearDown(): void
 
     public function testDocumentSnapshot(): void
     {
+        Feature::skipTestIfInActive('v6.8.0.0', $this);
```
Comment (MalteJanz) [inline tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: Do we really want to skip this whole snapshot test until the next major?

Comment (DennisGarding) [reply tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: no, I have add a switch and snapshots for 6.8. major now..

Comment (MalteJanz) [reply tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: Just curious, what exactly changes between them, or ask differently: why do we need a different snapshot for 6.8? Was it running into the edge case where the condition diverge now? 🤔

Comment (DennisGarding) [reply tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: No... it's just a whitespace issue... that's why my first solution was ugly. That creates no difference in white spaces between the HTML tags. 

This is not a displaying issue. Just the whitespace between the HTML tags. Maybe we should think about an adjustment in the snapshot comparison function. Maybe trim all whitespaces between the HTML tags.  :shrug:

Comment (MalteJanz) [reply tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: Some parts of HTML can be whitespace sensitive, thats why I would vote against trimming it for the snapshot comparision.

> No... it's just a whitespace issue

In that case I suggest we stick to a single snapshot and try to make template adjustments so the whitespace stays equal. That way it also serves as a proof that at least for the test case the output didn't change between 6.7 and upcoming 6.8, which is our overall goal to keep the same output during the refactor

Comment (DennisGarding) [reply tests/integration/Core/Checkout/Document/Renderer/InvoiceRendererTest.php]: Yes, thats the goal I currently work on. But until now it looks like it we need to make the conditions inline, like in my first throw

**Review by MalteJanz — APPROVED** [review]

File: src/Core/Framework/Resources/views/documents/includes/position_header.html.twig [hunk 6]
```diff
@@ -33,13 +34,19 @@ All blocks of this template are available in the template which renders this tem
...
                         {% set taxStatusGross = order.price.taxStatus == constant('Shopware\\Core\\Checkout\\Cart\\Price\\Struct\\CartPrice::TAX_STATE_GROSS') %}
                         {% set taxStatusNet = order.price.taxStatus == constant('Shopware\\Core\\Checkout\\Cart\\Price\\Struct\\CartPrice::TAX_STATE_NET') %}
 
-                        {% set displayVAT = not companyTaxEnabled or not (displayAdditionalNoteDelivery and isDeliveryCountry) %}
+                        {%- if not feature('v6.8.0.0') -%}
+                            {#- @deprecated tag:v6.8.0 - companyTaxEnabled will be removed without replace -#}
+                            {%- set companyTaxEnabled = billingAddress.country.companyTax.enabled -%}
+                            {#- @deprecated tag:v6.8.0 - displayAdditionalNoteDelivery will be removed without replace -#}
+                            {%- set displayAdditionalNoteDelivery = config.displayAdditionalNoteDelivery -%}
+                            {#- @deprecated tag:v6.8.0 - isDeliveryCountry will be removed without replace -#}
```
Comment (vintagesucks) [inline src/Core/Framework/Resources/views/documents/includes/position_header.html.twig]:
```suggestion
                            {#- @deprecated tag:v6.8.0 - companyTaxEnabled will be removed without replacement -#}
                            {%- set companyTaxEnabled = billingAddress.country.companyTax.enabled -%}
                            {#- @deprecated tag:v6.8.0 - displayAdditionalNoteDelivery will be removed without replacement -#}
                            {%- set displayAdditionalNoteDelivery = config.displayAdditionalNoteDelivery -%}
                            {#- @deprecated tag:v6.8.0 - isDeliveryCountry will be removed without replacement -#}
```

File: src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig [hunk 7]
```diff
@@ -28,9 +29,19 @@ All blocks of this template are available in the template which renders this tem
...
-                {% if (customer.customer.vatIds|first and config.displayCustomerVatId) or (customer.customer.vatIds|first and billingAddress.country.companyTax.enabled and config.displayAdditionalNoteDelivery and billingAddress.country.id in config.deliveryCountries) %}
-                    {{ 'document.vatId'|trans({'%vatId%': customer.customer.vatIds|first})|sw_sanitize(null, true) }}<br>
-                {% endif %}
+                {% set customerVatId = customer.customer.vatIds|first -%}
+                {% set displayCustomerVatId = customerVatId and (config.displayCustomerVatId or config.intraCommunityDelivery) -%}
+
+                {%- if not feature('v6.8.0.0') -%}
+                    {%- set customerVatId = customer.customer.vatIds|first -%}
+                    {%- set displayCustomerVatId = customerVatId and config.displayCustomerVatId -%}
+                    {#- @deprecated tag:v6.8.0 - displayCustomerVatIdForDelivery will be removed without replace -#}
```
Comment (vintagesucks) [inline src/Core/Framework/Resources/views/documents/includes/letter_header.html.twig]:
```suggestion
                    {#- @deprecated tag:v6.8.0 - displayCustomerVatIdForDelivery will be removed without replacement -#}
```

**Review by vintagesucks — APPROVED** [review]

**Review by larskemper — APPROVED** [review]

**Review by CR0YD — APPROVED** [review]
